WordPress/gutenberg
# ESLint: Stop disabling `react-hooks/exhaustive-deps` rule

## What?
Un-ignore all `react-hooks/exhaustive-deps` in the `@wordpress/components` package, and remove the custom rule that was setting severity of `react-hooks/exhaustive-deps` to be `error`. 

## Why?
Any `eslint-disable` would prevent the React Compiler from working on a specific file. However, we'd like to use React Compiler for as many files as possible (see https://github.com/WordPress/gutenberg/pull/61788 where we're experimenting with it). 

This came up recently in https://github.com/WordPress/gutenberg/pull/66207#discussion_r1806171696.

## How?
- Remove all `eslint-disable` directives for `react-hooks/exhaustive-deps` in `@wordpress/components`.
- Remove the custom ESLint rule that was changing `react-hooks/exhaustive-deps` severity to `error`. 

## Testing Instructions
Take a look at the lint check, part of the static analysis:
- You should see warnings for all files we're touching
- You should see no ESLint errors.

### Testing Instructions for Keyboard
Same

## Screenshots or screencast 
None
